### DISCUSSION TITLE
Plugin Authoring

### DISCUSSION THREAD

**Discussion by alijaya** [cocopon/tweakpane#426] [General]
Hello, thank you for making Tweakpane. I like to use it :D.
But when I want to try to create my own plugin, I'm stumped... there's the plugin template but it seems pretty complicated without documentation. Because it seems I need to understand first how tweakpane works and the architecture of it, and there's no documentation about that except dive down to the source code itself.

Anyone could give me direction where I should start?

**Comment by cocopon** [comment]
First of all, there are three types of [blades](https://cocopon.github.io/tweakpane/quick-tour/#blades). What is your choice?

- **Input bindings**
  User can change the value with your blade
- **Monitor bindings**
  User can monitor value change with your blade, but cannot change the bound value
- **Misc**
  Just provides a UI without binding

**Reply by alijaya** [reply]
Let's see... what if I want to make the Input Blades, how does it work?

For example, how tweakpane know which ui presented to the user? Because I've seen several number input with different kind of ui.
And how the binding works?

**Reply by cocopon** [reply]
Set `type: 'input'` to create a plugin that provides an input binding:
https://github.com/tweakpane/plugin-template/blob/a21333f2defd12e6cdc2a749c5e792fcad57f6d3/src/plugin.ts#L38

Each plugin has `accept` function and it decides whether this plugin should handle inputs or not:
https://github.com/tweakpane/plugin-template/blob/a21333f2defd12e6cdc2a749c5e792fcad57f6d3/src/plugin.ts#L44

In detail, when user runs `pane.addInput()`, Tweakpane checks a passed value and options with `accept` of registered plugins in order. If some plugin accepts an input value and options, Tweakpane creates a binding with its `binding` and a UI with its `controller` function.

**Reply by alijaya** [reply]
for the `accept` what is the priority here if there's two plugins could handle the same input value?

**Reply by cocopon** [reply]
The later one has higher priority. For example:

```js
pane.registerPlugin(A);
pane.registerPlugin(B);
```

In this case, `B.accept` is called first. And if `B` doesn't accept user inputs, then `A.accept` is called. If `A` also doesn't accept it, Tweakpane checks it with built-in plugins that have been registered at initialization.

Therefore, it's important not to accept wider input type. Checking `view: 'foobar'` option is useful for this purpose.
https://github.com/tweakpane/plugin-template/blob/a21333f2defd12e6cdc2a749c5e792fcad57f6d3/src/plugin.ts#L54

**Reply by alijaya** [reply]
Ah I see thanks!

And could you explain what `parseParams<PluginInputParams>` do?

**Reply by cocopon** [reply]
`parseParams` is literally a function that parse parameters, and returns a typed parameters object if it matches with the passed type. For example:

```ts
const result = parseParams<PluginInputParams>(params, {
  view: p.required.constant('dots'),

  max: p.optional.number,
  min: p.optional.number,
  step: p.optional.number,
});
```
https://github.com/tweakpane/plugin-template/blob/a21333f2defd12e6cdc2a749c5e792fcad57f6d3/src/plugin.ts#L52-L59

This means that params should have one required property and three optional properties.

```ts
// matched
pane.addInput(params, 'foobar', {
  view: 'dots',
});
pane.addInput(params, 'foobar', {
  view: 'dots',
  max: 10,
  min: 1,
  step: 1,
});

// not matched: required params `view` not found
pane.addInput(params, 'foobar', {
  max: 10,
  min: 1,
});

// not matched: value of `view` doesn't match
pane.addInput(params, 'foobar', {
  view: 'hello',
});

// not matched: type of `max` is different
pane.addInput(params, 'foobar', {
  view: 'dots',
  max: 'some',
});
```

This function is useful for identifying parameter types, but of course you can do it without this function.

**Reply by alijaya** [reply]
ah I see... I guess if it doesn't match it will return null?

**Reply by alijaya** [reply]
I think I have good idea what `accept` function does.
Now onto next part, the `binding` part, could you explain it to me?
It has 3 function: `reader`, `constraint`, and `writer`. What each function do? What is the arguments and what will it return?

**Reply by cocopon** [reply]
Binding is used to bind an external value (= parameter value, outside Tweakpane) and an internal value (inside Tweakpane).

For example, if user wants to add an input binding for color string `#ff0000`, they can write a code as below:

```ts
const params = {
 bg: '#ff0000',
};

pane.addInput(params, 'bg');
```

The external value `'#ff0000'` is inconvenient to process color data because it's a pain to extract each color component. This is why Tweakpane plugins have `binding` property.

In this complex case, Tweakpane creates an instance of internal [`Color`](https://github.com/cocopon/tweakpane/blob/66dd0b4ef7997cdcde826c8ae5f3c1fd2862ae2f/packages/core/src/input-binding/color/model/color.ts#L36) class with `binding.reader`. On the other hand, `binding.writer` is used to apply internal value changes to the bound value.

To sum up, `binding.reader` creates an internal value from the parameter value, and `binding.writer` applies the internal value to the bound value.

If you want to make a plugin that handle primitive values (e.g. number, string, boolean), its code can be simpler because the internal value and the external value is same type and you don't need to convert them.

`@tweakpane/plugin-template` is a simple binding example that treats number value:
https://github.com/tweakpane/plugin-template/blob/a21333f2defd12e6cdc2a749c5e792fcad57f6d3/src/plugin.ts#L71

**Reply by alijaya** [reply]
so to recap:
`reader` needs to return a function that convert user value to internal value
and 
`writer` needs to return a function that convert internal value to user value and set it to the `BindingTarget`?

**Reply by cocopon** [reply]
Right!

**Reply by alijaya** [reply]
I see... more or less, I think I understand about it...
Now about `controller`, what is it? how it works?

**Reply by cocopon** [reply]
Tweakpane adopts good old MVC architecture. `controller` function creates a controller and a view for user input.

`args.value` is a model that wraps an internal value created by `binding.reader`.
https://github.com/tweakpane/plugin-template/blob/a21333f2defd12e6cdc2a749c5e792fcad57f6d3/src/plugin.ts#L107

What you should do here is creating a controller for this model.

In this template plugin, controller creates a view, binds it to the model,
https://github.com/tweakpane/plugin-template/blob/a21333f2defd12e6cdc2a749c5e792fcad57f6d3/src/controller.ts#L37-L40

...handle UI events and apply it to the model.
https://github.com/tweakpane/plugin-template/blob/a21333f2defd12e6cdc2a749c5e792fcad57f6d3/src/controller.ts#L59

BTW are you using TypeScript for development? Type annotation can be helpful for code reading.

**Reply by alijaya** [reply]
Hmmm... I see... I guess I need to tinker and experiment it by myself...
Yes I'm using Typescript for development :D

**Comment by cocopon** [comment]
Created a guide for plugin development.
https://tweakpane.github.io/docs/plugins/dev/

Hope it helps...

**Reply by alijaya** [reply]
Hey thanks! That's looking good 😁
